### PR TITLE
CONSOLE-2393: i18n user management page

### PR DESCRIPTION
**Pull request by wansc2016**
The PR includes i18n support for the following pages:
User management
- Users
- Groups
- Service Accounts
- Roles
- Role Bindings

[CONSOLE-2393](https://issues.redhat.com/browse/CONSOLE-2393)

### CODE REVIEW

File: frontend/public/components/RBAC/bindings.jsx [hunk 1]
```diff
@@ -291,27 +286,50 @@ export const RoleBindingsPage = ({
...
+      flatten={flatten}
+      label={t('bindings~Role Bindings')}
+      ListComponent={BindingsList}
+      namespace={namespace}
+      resources={roleResources}
+      rowFilters={staticFilters ? [] : rowFilters}
+      staticFilters={staticFilters}
+      showTitle={showTitle}
+      textFilter="role-binding"
+      title={t('bindings~Role Bindings')}
```
Comment (spadgett) [inline frontend/public/components/RBAC/bindings.jsx:329]:
```suggestion
      title={t('bindings~RoleBindings')}
```

File: frontend/public/components/RBAC/bindings.jsx [hunk 2]
```diff
@@ -331,47 +349,32 @@ const NsRoleDropdown_ = (props) => {
...
   const resourceForKind = (kind) => ({ kind, namespace: kind === 'Role' ? props.namespace : null });
   const resources = _.map(kinds, resourceForKind);
+  const { t } = props;
 
   return (
     <ListDropdown
       {...props}
       dataFilter={roleFilter}
-      desc="Namespace Roles (Role)"
+      desc={t('bindings~Namespace roles (role)')}
```
Comment (spadgett) [inline frontend/public/components/RBAC/bindings.jsx:358]:
```suggestion
      desc={t('bindings~Namespace roles (Role)')}
```

File: frontend/public/components/RBAC/bindings.jsx [hunk 3]
```diff
@@ -381,248 +384,281 @@ const Section = ({ label, children }) => (
...
+            desc: t(
+              'bindings~Grant the permissions to a user or set of users at the cluster level and in all namespaces.',
+            ),
+          },
+        ];
+
+        const subjectKinds = [
+          { value: 'User', title: t('bindings~User') },
+          { value: 'Group', title: t('bindings~Group') },
+          { value: 'ServiceAccount', title: t('bindings~Service account') },
```
Comment (spadgett) [inline frontend/public/components/RBAC/bindings.jsx:530]:
```suggestion
          { value: 'ServiceAccount', title: t('bindings~ServiceAccount') },
```

File: frontend/public/components/RBAC/bindings.jsx [hunk 4]
```diff
@@ -381,248 +384,281 @@ const Section = ({ label, children }) => (
...
+
+              {!_.get(fixed, 'kind') && (
+                <Section label={t('bindings~Binding type')}>
+                  <RadioGroup currentValue={kind} items={bindingKinds} onChange={this.setKind} />
+                </Section>
+              )}
+
+              <div className="co-form-section__separator" />
+
+              <Section label={t('bindings~Role Binding')}>
```
Comment (spadgett) [inline frontend/public/components/RBAC/bindings.jsx:554]:
```suggestion
              <Section label={t('bindings~RoleBinding')}>
```

File: frontend/public/components/RBAC/bindings.jsx [hunk 5]
```diff
@@ -381,248 +384,281 @@ const Section = ({ label, children }) => (
...
-                    onChange={this.changeNamespace}
-                    id="ns-dropdown"
+                  {_.get(fixed, 'metadata.name') ? (
+                    <ResourceName kind={kind} name={metadata.name} />
+                  ) : (
+                    <input
+                      className="pf-c-form-control"
+                      type="text"
+                      onChange={this.changeName}
+                      placeholder={t('bindings~Role binding name')}
```
Comment (spadgett) [inline frontend/public/components/RBAC/bindings.jsx:566]:
```suggestion
                      placeholder={t('bindings~RoleBinding name')}
```

File: frontend/public/components/RBAC/role.jsx [hunk 6]
```diff
@@ -31,7 +33,7 @@ const menuActions = [
   //   href: addHref(role.metadata.name, role.metadata.namespace),
   // }),
   (kind, role) => ({
-    label: 'Add Role Binding',
+    label: i18next.t('role~Add role binding'),
```
Comment (spadgett) [inline frontend/public/components/RBAC/role.jsx:36]:
```suggestion
    label: i18next.t('role~Add RoleBinding'),
```

File: frontend/public/components/RBAC/role.jsx [hunk 7]
```diff
@@ -236,44 +227,72 @@ export const BindingsForRolePage = (props) => {
...
-);
+export const RolesDetailsPage = (props) => {
+  const { t } = useTranslation();
+  return (
+    <DetailsPage
+      {...props}
+      pages={[
+        navFactory.details(DetailsWithTranslate),
+        navFactory.editYaml(),
+        { href: 'bindings', name: t('role~Role bindings'), component: BindingsForRolePage },
```
Comment (spadgett) [inline frontend/public/components/RBAC/role.jsx:245]:
```suggestion
        { href: 'bindings', name: t('role~RoleBindings'), component: BindingsForRolePage },
```

File: frontend/public/components/RBAC/role.jsx [hunk 8]
```diff
@@ -236,44 +227,72 @@ export const BindingsForRolePage = (props) => {
...
-  <MsgBox
-    title="No Roles Found"
-    detail="Roles grant access to types of objects in the cluster. Roles are applied to a team or user via a Role Binding."
-  />
-);
+const EmptyMsg = () => {
+  const { t } = useTranslation();
+  return (
+    <MsgBox
+      title={t('role~No roles found')}
```
Comment (spadgett) [inline frontend/public/components/RBAC/role.jsx:258]:
```suggestion
      title={t('role~No Roles found')}
```

File: frontend/public/components/RBAC/role.jsx [hunk 9]
```diff
@@ -236,44 +227,72 @@ export const BindingsForRolePage = (props) => {
...
+      detail={t(
+        'role~Roles grant access to types of objects in the cluster. Roles are applied to a team or user via a Role Binding.',
```
Comment (spadgett) [inline frontend/public/components/RBAC/role.jsx:260]:
```suggestion
        'role~Roles grant access to types of objects in the cluster. Roles are applied to a team or user via a RoleBinding.',
```

File: frontend/public/components/RBAC/role.jsx [hunk 10]
```diff
@@ -291,14 +310,15 @@ export const RolesPage = ({ namespace, mock, showTitle }) => {
...
+  const { t } = useTranslation();
   return (
     <MultiListPage
       ListComponent={RolesList}
       canCreate={true}
       showTitle={showTitle}
       namespace={namespace}
       createAccessReview={accessReview}
-      createButtonText="Create Role"
+      createButtonText={t('role~Create role')}
```
Comment (spadgett) [inline frontend/public/components/RBAC/role.jsx:321]:
```suggestion
      createButtonText={t('role~Create Role')}
```

File: frontend/public/components/group.tsx [hunk 11]
```diff
@@ -22,9 +22,11 @@ import {
   SectionHeading,
   Timestamp,
 } from './utils';
+import { useTranslation } from 'react-i18next';
+import i18next from 'i18next';
 
 const addUsers: KebabAction = (kind: K8sKind, group: GroupKind) => ({
-  label: 'Add Users',
+  label: i18next.t('group~Add users'),
```
Comment (spadgett) [inline frontend/public/components/group.tsx:29]:
```suggestion
  label: i18next.t('group~Add Users'),
```

File: frontend/public/components/group.tsx [hunk 12]
```diff
@@ -34,7 +36,7 @@ const addUsers: KebabAction = (kind: K8sKind, group: GroupKind) => ({
 
 const removeUser = (group: GroupKind, user: string): KebabOption => {
   return {
-    label: 'Remove User',
+    label: i18next.t('group~Remove user'),
```
Comment (spadgett) [inline frontend/public/components/group.tsx:39]:
```suggestion
    label: i18next.t('group~Remove User'),
```

File: frontend/public/components/modals/add-users-modal.tsx [hunk 13]
```diff
@@ -23,18 +24,24 @@ export const AddUsersModal = withHandlePromise((props: AddUsersModalProps) => {
...
+  const { t } = useTranslation();
 
   return (
     <form onSubmit={submit} name="form" className="modal-content ">
-      <ModalTitle>Add Users</ModalTitle>
+      <ModalTitle>{t('add-users-modal~Add users')}</ModalTitle>
       <ModalBody>
-        <p>Add new users to group {props.group.metadata.name}.</p>
-        <ListInput label="Users" required initialValues={values} onChange={setValues} />
+        <p>{t('add-users-modal~Add new users to group {name}.', props.group.metadata)}</p>
```
Comment (spadgett) [inline frontend/public/components/modals/add-users-modal.tsx:33]:
```suggestion
        <p>{t('add-users-modal~Add new Users to Group {name}.', props.group.metadata)}</p>
```

File: frontend/public/components/modals/remove-user-modal.tsx [hunk 14]
```diff
@@ -21,18 +22,23 @@ export const RemoveUserModal = withHandlePromise((props: RemoveUserModalProps) =
     return props.handlePromise(k8sPatch(GroupModel, props.group, patch), props.close);
   };
 
+  const { t } = useTranslation();
   return (
     <form onSubmit={submit} name="form" className="modal-content ">
       <ModalTitle>
-        <YellowExclamationTriangleIcon className="co-icon-space-r" /> Remove User from Group?
+        <YellowExclamationTriangleIcon className="co-icon-space-r" />
+        {t('remove-user-modal~Remove user from group?')}
```
Comment (spadgett) [inline frontend/public/components/modals/remove-user-modal.tsx:30]:
```suggestion
        {t('remove-user-modal~Remove User from Group?')}
```

File: frontend/public/components/modals/remove-user-modal.tsx [hunk 15]
```diff
@@ -21,18 +22,23 @@ export const RemoveUserModal = withHandlePromise((props: RemoveUserModalProps) =
...
       </ModalTitle>
       <ModalBody>
-        Remove user {props.user} from group {props.group.metadata.name}?
+        {t('remove-user-modal~Remove user {user} from group {name}?', {
```
Comment (spadgett) [inline frontend/public/components/modals/remove-user-modal.tsx:33]:
```suggestion
        {t('remove-user-modal~Remove User {user} from Group {name}?', {
```

File: frontend/public/components/service-account.jsx [hunk 16]
```diff
@@ -41,7 +43,9 @@ const KubeConfigify = (kind, sa) => ({
             saSecretsByName[s.metadata.name] && s.type === 'kubernetes.io/service-account-token',
         );
         if (!secret) {
-          errorModal({ error: 'Unable to get service account token.' });
+          errorModal({
+            error: i18next.t('service-account~Unable to get service account token.'),
```
Comment (spadgett) [inline frontend/public/components/service-account.jsx:47]:
```suggestion
            error: i18next.t('service-account~Unable to get ServiceAccount token.'),
```

File: frontend/public/components/service-account.jsx [hunk 17]
```diff
@@ -181,19 +150,20 @@ const Details = ({ obj: serviceaccount }) => {
     secrets,
   } = serviceaccount;
   const filters = { selector: { field: 'metadata.name', values: new Set(_.map(secrets, 'name')) } };
+  const { t } = useTranslation();
 
   return (
     <>
       <div className="co-m-pane__body">
-        <SectionHeading text="Service Account Details" />
+        <SectionHeading text={t('service-account~Service Account details')} />
```
Comment (spadgett) [inline frontend/public/components/service-account.jsx:158]:
```suggestion
        <SectionHeading text={t('service-account~ServiceAccount details')} />
```

File: frontend/public/components/user.tsx [hunk 18]
```diff
@@ -98,42 +73,90 @@ const UserTableRow: RowFunction<UserKind> = ({ obj, index, key, style }) => {
   );
 };
 
-const EmptyMsg = () => <MsgBox title="No Users Found" />;
+const EmptyMsg = () => {
+  const { t } = useTranslation();
+  return <MsgBox title={t('user~No users found')} />;
```
Comment (spadgett) [inline frontend/public/components/user.tsx:78]:
```suggestion
  return <MsgBox title={t('user~No Users found')} />;
```

File: frontend/public/components/user.tsx [hunk 19]
```diff
@@ -98,42 +73,90 @@ const UserTableRow: RowFunction<UserKind> = ({ obj, index, key, style }) => {
...
-    kind={referenceForModel(UserModel)}
-    ListComponent={UserList}
-    canCreate={false}
-  />
-);
+const NoDataEmptyMsg = () => {
+  const { t } = useTranslation();
+  return (
+    <MsgBox
+      title={t('user~No users found')}
```
Comment (spadgett) [inline frontend/public/components/user.tsx:86]:
```suggestion
      title={t('user~No Users found')}
```

**Review by spadgett — COMMENTED** [review]
Thank you for the PR!

I apologize for not reviewing sooner. I had the comments finished, but forgot to submit the review. Please feel free to reach out if you need a code review. Thanks.

Looks good overall. We are switching to use these conventions for capitalization as we externalize the strings: http://openshift.github.io/openshift-origin-design/conventions/documentation/capitalization.html

File: frontend/public/components/RBAC/bindings.jsx [hunk 20]
```diff
@@ -64,12 +66,12 @@ export const flatten = (resources) =>
     return ret;
   });
 
-const menuActions = ({ subjectIndex, subjects }, startImpersonate) => {
+const menuActions = ({ subjectIndex, subjects }, startImpersonate_) => {
```
[the remark this answers is not in the document]
Comment (wansc2016) [reply frontend/public/components/RBAC/bindings.jsx:69]: Got it. I will revoke it back.

**Comment by spadgett** [comment]
Some build errors:

```
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx(513,109):
TS1002: Unterminated string literal.

ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx(514,13):
TS1005: ',' expected.

ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx(514,18):
TS1005: ',' expected.

ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx(514,22):
TS1005: ',' expected.

ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx(514,30):
TS1003: Identifier expected.

ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx
ERROR in /go/src/github.com/openshift/console/frontend/public/components/RBAC/bindings.jsx(514,31):
TS1002: Unterminated string literal.
```

File: frontend/public/components/RBAC/bindings.jsx [hunk 21]
```diff
@@ -222,7 +227,7 @@ const RoleBindingsTableRow = ({ obj: binding, index, key, style }) => {
         {binding.metadata.namespace ? (
           <ResourceLink kind="Namespace" name={binding.metadata.namespace} />
         ) : (
-          'All Namespaces'
+          i18next.t('bindings~All Namespaces')
```
[the remark this answers is not in the document]
Comment (wansc2016) [reply frontend/public/components/RBAC/bindings.jsx]: When I changed it with useTranslation, the React complains that 
"Invalid hook call. Hooks can only be called inside of the body of a function component."
It looks like RowFunction is a function, other than a function component. So I had to change it back to i18next.t

File: frontend/public/components/RBAC/role.jsx [hunk 22]
```diff
@@ -202,15 +161,46 @@ const BindingsTableRow = ({ obj: binding, index, key, style }) => {
       <TableData className={bindingsColumnClasses[1]}>{binding.subject.kind}</TableData>
       <TableData className={bindingsColumnClasses[2]}>{binding.subject.name}</TableData>
       <TableData className={bindingsColumnClasses[3]}>
-        {binding.namespace || 'All Namespaces'}
+        {binding.namespace || i18next.t('role~All Namespaces')}
```
[the remark this answers is not in the document]
Comment (wansc2016) [reply frontend/public/components/RBAC/role.jsx]: When I changed it with useTranslation, I got the same result with the one in bindings.tsx. React complains that 
"Invalid hook call. Hooks can only be called inside of the body of a function component."
It looks like RowFunction is a function, other than a function component. So I had to change it back to i18next.t

**Comment by wansc2016** [comment]
@spadgett Should we change Role Binding to RoleBinding in all places? Will it be inconsistent with Navigation Menu?

**Comment by spadgett** [comment]
> @spadgett Should we change Role Binding to RoleBinding in all places? Will it be inconsistent with Navigation Menu?

Yes, #7277 updates the navigation menu.

**Review by spadgett — COMMENTED** [review]
LGTM, except it looks like we have a test failure on the role bindings page. This is ready for testing and approvals, however.

/assign @yapei @ahardin-rh @reestr

**Comment by yanpzhan** [comment]
Reported a bug about the pr: https://bugzilla.redhat.com/show_bug.cgi?id=1915723

**Comment by yanpzhan** [comment]
"Add more" is not marked for translation on add user model:
Screenshot from 2021-01-13 16-27-35

**Comment by wansc2016** [comment]
@yanpzhan https://bugzilla.redhat.com/show_bug.cgi?id=1915723 has been fixed.

"Add more" issue will be fixed with https://issues.redhat.com/browse/CONSOLE-2474

**Review by spadgett — APPROVED** [review]
/lgtm
/hold for docs approval

@ahardin-rh PTAL, thanks!

**Comment by spadgett** [comment]
Thanks @wansc2016 for helping us with this!

**Pull request merged**